### PR TITLE
feat(spring): use shared configuration

### PR DESCRIPTION
**Pull request by diegomarquezp**
Modification of generated autoconfig to use shared properties, as demonstrated [here](https://github.com/zhumin8/language_poc1/blob/multi-module-w-shared/cloud-language-starter/src/main/java/com/sample/autoconfig/LanguageAutoConfig.java)

It will have changes coming from [`autoconfig-gen-draft2-array-annotation`](https://github.com/googleapis/gapic-generator-java/pull/1070) (should be reviewed first) since it depends on anonymous arrays for annotations. 
The main changes are in the autoconfig composer and updated golden files.

### CODE REVIEW

**Comment by diegomarquezp** [comment]
Force pushed a rebase on [`autoconfig-gen-draf2-array-annotation`](https://github.com/googleapis/gapic-generator-java/pull/1070)

**Comment by diegomarquezp** [comment]
Another force push to match current changes of `autoconfig-gen-draft2-array-annotation`

File: src/main/java/com/google/api/generator/spring/composer/SpringAutoConfigClassComposer.java [hunk 1]
```diff
@@ -155,41 +158,91 @@ private static List<Statement> createMemberVariables(
...
+    //             .setName("coreCredentialsProvider")
+    //             .setType(types.get("CredentialsProvider"))
+    //             .build());
+    //
+    // VariableExpr coreProjectIdProviderVarExpr =
+    //     VariableExpr.withVariable(
+    //         Variable.builder()
+    //             .setName("coreProjectIdProvider")
+    //             .setType(types.get("GcpProjectIdProvider"))
+    //             .build());
```
Comment (zhumin8) [inline src/main/java/com/google/api/generator/spring/composer/SpringAutoConfigClassComposer.java:181]: The commented out code can be deleted. It probably sneaked in from rebase/merge.

File: src/main/java/com/google/api/generator/spring/composer/SpringAutoConfigClassComposer.java [hunk 2]
```diff
@@ -155,41 +158,91 @@ private static List<Statement> createMemberVariables(
...
+        VariableExpr.withVariable(
+            Variable.builder()
+                .setName("globalProperties")
+                .setType(types.get("GlobalProperties"))
+                .build());
+    // Variable projectIdProviderVar =
+    //     Variable.builder()
+    //         .setName("projectIdProvider")
+    //         .setType(types.get("GcpProjectIdProvider"))
+    //         .build();
```
Comment (zhumin8) [inline src/main/java/com/google/api/generator/spring/composer/SpringAutoConfigClassComposer.java:199]: same here.

File: src/main/java/com/google/api/generator/spring/utils/GlobalPropertiesUtils.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import com.google.api.generator.engine.ast.ScopeNode;
+import com.google.api.generator.engine.ast.Statement;
+import com.google.api.generator.engine.ast.TypeNode;
+import com.google.api.generator.engine.ast.VaporReference;
+import com.google.api.generator.engine.ast.Variable;
+import com.google.api.generator.engine.ast.VariableExpr;
+import java.util.Map;
+
+public class GlobalPropertiesUtils {
+  public static final String GLOBAL_PROPERTIES_CLAZZ_NAME = "GlobalProperties";
```
Comment (zhumin8) [inline src/main/java/com/google/api/generator/spring/utils/GlobalPropertiesUtils.java:27]: Dropping a reminder note here to change the corresponding classes in https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/1308

**Review by zhumin8 — APPROVED** [review]
LGTM

File: src/main/java/com/google/api/generator/spring/composer/SpringAutoConfigClassComposer.java (hunk 1 above)
Comment (diegomarquezp) [reply src/main/java/com/google/api/generator/spring/composer/SpringAutoConfigClassComposer.java]: Removed!

File: src/main/java/com/google/api/generator/spring/utils/GlobalPropertiesUtils.java (hunk 3 above)
Comment (diegomarquezp) [reply src/main/java/com/google/api/generator/spring/utils/GlobalPropertiesUtils.java:27]: Removed!

File: src/main/java/com/google/api/generator/spring/composer/SpringAutoConfigClassComposer.java (hunk 2 above)
Comment (diegomarquezp) [reply src/main/java/com/google/api/generator/spring/composer/SpringAutoConfigClassComposer.java]: Removed!

File: src/main/java/com/google/api/generator/spring/composer/SpringAutoConfigClassComposer.java [hunk 4]
```diff
@@ -955,6 +1061,7 @@ private static Map<String, TypeNode> createDynamicTypes(Service service, String
     typeMap.put("ServiceClient", serviceClient);
     typeMap.put("ServiceSettings", serviceSettings);
     typeMap.put("ServiceSettingsBuilder", serviceSettingsBuilder);
+    typeMap.put("GlobalProperties", GlobalPropertiesUtils.getGlobalPropertiesType());
```
Comment (diegomarquezp) [inline src/main/java/com/google/api/generator/spring/composer/SpringAutoConfigClassComposer.java:1064]: We will need to add a concrete reference once we merge https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/1308

Comment (diegomarquezp) [reply src/main/java/com/google/api/generator/spring/composer/SpringAutoConfigClassComposer.java:1064]: Solved as we have full control of the global properties in spring-cloud-gcp

File: src/test/java/com/google/api/generator/spring/composer/goldens/EchoSpringAutoConfigurationFull.golden [hunk 5]
```diff
@@ -76,7 +81,10 @@ public class EchoSpringAutoConfiguration {
   @Bean
   @ConditionalOnMissingBean
   public CredentialsProvider echoCredentials() throws IOException {
-    return ((CredentialsProvider) new DefaultCredentialsProvider(this.clientProperties));
+    if (this.clientProperties.getCredentials().hasKey()) {
+      return ((CredentialsProvider) new DefaultCredentialsProvider(this.clientProperties));
+    }
+    return ((CredentialsProvider) new DefaultCredentialsProvider(this.GlobalProperties));
```
Comment (emmileaf) [inline src/test/java/com/google/api/generator/spring/composer/goldens/EchoSpringAutoConfigurationFull.golden:87]: Access to field (lowerCamel) here?
```suggestion
    return ((CredentialsProvider) new DefaultCredentialsProvider(this.globalProperties));
```

File: src/test/java/com/google/api/generator/spring/composer/goldens/EchoSpringAutoConfigurationFull.golden [hunk 6]
```diff
@@ -86,7 +94,7 @@ public class EchoSpringAutoConfiguration {
   @Bean
   @ConditionalOnMissingBean
   public TransportChannelProvider defaultEchoTransportChannelProvider() {
-    return EchoSettings.defaultTransportChannelProvider();
+    return defaultTransportChannelProvider();
```
Comment (emmileaf) [inline src/test/java/com/google/api/generator/spring/composer/goldens/EchoSpringAutoConfigurationFull.golden:97]: Is this change intentional? (looks like a self-reference now)

**Review by emmileaf — COMMENTED** [review]
Left comments on one of the golden files so it's easier to read, but please resolve in corresponding composer/util methods - thanks!

File: src/test/java/com/google/api/generator/spring/composer/goldens/EchoSpringAutoConfigurationFull.golden (hunk 6 above)
Comment (diegomarquezp) [reply src/test/java/com/google/api/generator/spring/composer/goldens/EchoSpringAutoConfigurationFull.golden:97]: Ah, it should be getting `EchoSettings` from `types` instead of `STATIC_TYPES` since it's a `VaporReference`. I'm pushing a fix

File: src/test/java/com/google/api/generator/spring/composer/goldens/EchoSpringAutoConfigurationFull.golden (hunk 5 above)
Comment (diegomarquezp) [reply src/test/java/com/google/api/generator/spring/composer/goldens/EchoSpringAutoConfigurationFull.golden:87]: Thanks for catching this!
